### PR TITLE
Ensure responses are valid before showing them #4486

### LINKED ISSUE #4486: Ensure responses are valid before showing them
In results, submit pages.
Related to #4482

### PR DESCRIPTION
**Pull request by Sumei1009**
fixes #4486

### CODE REVIEW

**Comment by damithc** [comment]
SLAP is pretty bad in some places. I noticed a few very long methods.

File: src/main/java/teammates/common/datatransfer/FeedbackConstantSumResponseDetails.java [hunk 1]
```diff
@@ -39,6 +39,10 @@ public void extractResponseDetails(FeedbackQuestionType questionType,
     @Override
     public String getAnswerString() {
         String listString = answers.toString();//[1, 2, 3] format
+        if (answers.get(0) == -1) {
+            listString = "Invalid response";
+            return listString;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackConstantSumResponseDetails.java]: See if you can change things like this to just `return "Invalid response"`

File: src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java [hunk 2]
```diff
@@ -65,27 +69,34 @@ public String getAnswerString() {
     public String getAnswerHtml(FeedbackQuestionDetails questionDetails) {
         FeedbackRankOptionsQuestionDetails rankQuestion = (FeedbackRankOptionsQuestionDetails) questionDetails;
         
-        SortedMap<Integer, List<String>> orderedOptions = generateMapOfRanksToOptions(rankQuestion);
-        
+        String invalidAnswer = "Invalid response";
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java]: This const seem to be used repeatedly. Extract it to the `Const` class?

File: src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java [hunk 3]
```diff
@@ -65,27 +69,34 @@ public String getAnswerString() {
...
         
-        for (Entry<Integer, List<String>> rankAndOption : orderedOptions.entrySet()) {
-            Integer rank = rankAndOption.getKey();
-            if (rank == Const.POINTS_NOT_SUBMITTED) {
-                continue;
-            }
+        if (answers.get(0) == -1) {
+            htmlBuilder.append("<li>");
+            htmlBuilder.append(invalidAnswer);
+            htmlBuilder.append("</li>");
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java]: `htmlBuilder.append("<li>" + invalidAnswer + "</li>");` will do the trick.

File: src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java [hunk 4]
```diff
@@ -65,27 +69,34 @@ public String getAnswerString() {
...
+                    continue;
+                }
+                
+                List<String> optionsWithGivenRank = rankAndOption.getValue();
+                for (String option : optionsWithGivenRank) {
+                    htmlBuilder.append("<li>");
+                    htmlBuilder.append(Sanitizer.sanitizeForHtml(rank.toString()));
+                    htmlBuilder.append(": ");
+                    htmlBuilder.append(option);
+                    htmlBuilder.append("</li>");
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackRankOptionsResponseDetails.java]: Here too. If appending everything together makes it too long, at least append some of them. You don't need to separate every single thing.

File: src/main/java/teammates/common/datatransfer/FeedbackContributionResponseDetails.java [hunk 5]
```diff
@@ -45,23 +45,35 @@ public int getAnswer() {
     
     @Override
     public String getAnswerString() {
+        if (answer == -1) {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackContributionResponseDetails.java]: Extract `-1` to a constant, something like `private static final int CONST_INVALID_RESPONSE = -1;`.

**Comment by kanghj** [comment]
Is showing "Invalid Response" really the right thing to do in this situation? The issue also talks about handling invalid responses for the submission pages, is that handled here?

File: src/main/java/teammates/common/datatransfer/FeedbackResponseAttributes.java [hunk 6]
```diff
@@ -218,5 +225,52 @@ public int compare(FeedbackResponseAttributes fr1, FeedbackResponseAttributes fr
...
+    public void setCorruptionMessage(String errorMessage) {
+        corruptedMessage = errorMessage;
+    }
+    
+    public String constructJsonCorruptedMessage(Class<? extends FeedbackResponseDetails> responseDetailsClass) {
+        if (responseDetailsClass == FeedbackTextResponseDetails.class) {
+            return corruptedMessage;
+        }
+        if (responseDetailsClass == FeedbackConstantSumResponseDetails.class) {
+            return "{\"questionType\" : \"CONTSUM\",\"answers\" : [-1] }";
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackResponseAttributes.java]: Even if we are going to do this, we should put things like this in the specific question type's ResponseDetails/QuestionDetails page.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 7]
```diff
@@ -121,13 +123,255 @@ public FeedbackSessionResultsBundle(FeedbackSessionAttributes feedbackSession,
...
+            case TEAMS:
+                if (!isGiverVisible(response)) {
+                    if (!response.giverEmail.contains("Anonymous team")) {
+                        errorMessage = Const.INVALID_GIVER_NOT_AN_ANONYMOUS_TEAM;
+                        return false;
+                    }
+                } else {
+                    if (isParticipantIdentifierStudent(response.giverEmail)) {
+                        giver = getTeamNameForEmail(response.giverEmail);
+                    } else if (isParticipantIdentifierInstructor(response.giverEmail)) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Was this copied from somewhere?

**Comment by thyageshm** [comment]
@Sumei1009, can this be completed by this release? if not, please push the milestone accordingly

File: src/main/java/teammates/common/datatransfer/FeedbackResponseAttributes.java (hunk 6 above)
Comment (Sumei1009) [inline src/main/java/teammates/common/datatransfer/FeedbackResponseAttributes.java]: The right format json data should be passed to responseDetails class, so I construct the json data here. Do you mean the invalid response message should be handled by ResponseDetails/QuestionDetails page?

**Comment by thyageshm** [comment]
@Sumei1009 please complete this by V5.72

**Comment by wkurniawan07** [comment]
@kanghj I have a feeling that this might not be the right way to do. Your thoughts?

**Comment by kanghj** [comment]
My general feeling is that the code for detecting invalid responses should be in the logic layer, rather than in the sessions bundle. There is no need to detect anything about visibility and anonymous giver/recipient, in my opinion. You may find that there are already methods in the logic layer to get the possible recipients/givers for each question, you can then tally that against the responses.

**Pull request closed without merging**